datasource/ethereum/src/ethereum_adapter.rs

```diff
@@ -255,39 +256,61 @@ where
                 return None;
             }
-            let end = start + step;
-            debug!(logger, "Requesting logs for blocks [{}, {}]", start, end);
+
+            // This will iterate more than once iff `step` has been reduced,
+            // making as many parallel requests of size `step` as necessary,
+            // respecting `LOG_STREAM_PARALLEL_CHUNKS`.
+            let mut chunk_futures = vec![];
+            let mut offset = start;
+            for _ in 0..*LOG_STREAM_PARALLEL_CHUNKS {
+                if offset == to + 1 {
+                    break;
+                }
+                let log_filter = log_filter.clone();
+                let next_offset = (offset + step).min(to) + 1;
+                debug!(
+                    logger,
+                    "Requesting logs for blocks [{}, {}]",
+                    offset,
+                    next_offset - 1
+                );
+                chunk_futures.push(
+                    eth.logs_with_sigs(
+                        &logger,
+                        offset,
+                        next_offset - 1,
+                        addresses.clone(),
+                        event_sigs.clone(),
+                        TOO_MANY_LOGS_FINGERPRINT,
+                    )
+                    .map(move |logs| {
+                        logs.into_iter()
+                            .filter(|log| log_filter.matches(log))
+                            .collect::<Vec<Log>>()
+                    }),
+                );
+                offset = next_offset;
+            }
             let logger = logger.clone();
-            let log_filter = log_filter.clone();
             Some(
-                eth.logs_with_sigs(
-                    &logger,
-                    start,
-                    end,
-                    addresses.clone(),
-                    event_sigs.clone(),
-                    TOO_MANY_LOGS_FINGERPRINT,
-                )
-                .map(move |logs| {
-                    logs.into_iter()
-                        .filter(move |log| log_filter.matches(log))
-                        .collect::<Vec<Log>>()
-                })
-                .then(move |res| match res {
-                    Err(e) => {
-                        let string_err = e.to_string();
-                        if string_err.contains(TOO_MANY_LOGS_FINGERPRINT) {
-                            let new_step = step / 10;
-                            debug!(logger, "Reducing block range size to scan for events";
+                stream::futures_ordered(chunk_futures)
+                    .collect()
+                    .map(|chunks| chunks.into_iter().flatten().collect::<Vec<Log>>())
+                    .then(move |res| match res {
+                        Err(e) => {
+                            let string_err = e.to_string();
+                            if string_err.contains(TOO_MANY_LOGS_FINGERPRINT) {
+                                let new_step = step / 10;
```

Comment:
Should we handle the lower bound here differently?

What if the step is already `1`? I think step is going to turn into `0` and then the stream `unfold` will never end.
Maybe we should just log and error in that case.

Reply:
`0` is fine, the range is inclusive and we always increase the start so it always advances by at least one block. This is worth a comment.

Reply:
Yeah, if we are coming from a step of `0` or `1` though we already know there are too many logs in that block for Infura to return a response. Would we just get stuck in a loop with a step size of `0` and the `TOO_MANY_LOGS_ERROR` on every fold? 

Might be worth logging (and erroring?) that the smallest step size is still too big.

Reply:
If we're coming from a step of `0` this will loop forever. I'll add a check for this.